lib/src/java/facter.cc

```diff
@@ -0,0 +1,169 @@
+    }
+    if (auto ptr = dynamic_cast<array_value const*>(val)) {
+        auto array = env->NewObjectArray(ptr->size(), object_class, nullptr);
+
+        // Recurse on each element of the array
+        jsize index = 0;
+        ptr->each([&](value const* element) {
+            env->SetObjectArrayElement(array, index++, to_object(env, element));
+            return true;
+        });
+        return array;
+    }
+    if (auto ptr = dynamic_cast<map_value const*>(val)) {
+        auto hashmap = env->NewObject(hash_class, hash_constructor, static_cast<jint>(ptr->size()));
+
+        // Recurse on each element in the map
+        ptr->each([&](string const& name, value const* element) {
+            env->CallObjectMethod(hashmap, hash_put, env->NewStringUTF(name.c_str()), to_object(env, element));
+            return true;
+        });
+        return hashmap;
+    }
+    return nullptr;
+}
+
+extern "C" {
+    LIBFACTER_EXPORT jint JNI_OnLoad(JavaVM* vm, void* reserved)
+    {
+        JNIEnv* env;
+        if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
+            return JNI_ERR;
+        }
+        // For the classes, we need global refs
+        object_class = static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Object")));
+        if (!object_class) {
+            return JNI_ERR;
+        }
+        long_class = static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Long")));
+        if (!long_class) {
+            return JNI_ERR;
+        }
+        double_class = static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Double")));
+        if (!double_class) {
+            return JNI_ERR;
+        }
+        boolean_class = static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Boolean")));
+        if (!boolean_class) {
+            return JNI_ERR;
+        }
+        hash_class = static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/util/HashMap")));
+        if (!hash_class) {
+            return JNI_ERR;
+        }
+
+        // For the method ids, we can keep these cached as is
+        long_constructor = env->GetMethodID(long_class, "<init>", "(J)V");
+        double_constructor = env->GetMethodID(double_class, "<init>", "(D)V");
+        boolean_constructor = env->GetMethodID(boolean_class, "<init>", "(Z)V");;
+        hash_constructor = env->GetMethodID(hash_class, "<init>", "(I)V");
+        hash_put = env->GetMethodID(hash_class, "put", "(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;");
+
+        setup_logging(boost::nowide::cerr);
+        set_level(level::warning);
+
+        facts_collection.reset(new collection());
+        facts_collection->add_default_facts();
+        facts_collection->add_external_facts();
+        return JNI_VERSION_1_6;
+    }
+
+    LIBFACTER_EXPORT void JNI_OnUnload(JavaVM* vm, void* reserved)
+    {
+        if (!facts_collection) {
+            return;
```

Comment:
Can `facts_collection` be nil (as we call `.reset` in other places without checking)? If it can be null, will we return too early before deleting the global refs?